### PR TITLE
fix(app-shell): to handle errors without extentions property

### PR DESCRIPTION
**Pull request by tdeekens**
#### Summary

This pull request fixes token regeneration for GraphQL APIs which do not have an `extentions` property.

#### Description

Some internal APIs do not expose an `extensions` property but an error which looks e.g. like this

```json
{"statusCode":401,"message":"invalid_token","errors":[{"code":"invalid_token","message":"invalid_token"}],"error":"invalid_token"}
```

As a result I propose that we don't only assume a GraphQL error based on the presence of the `extensions` property but also based on the `message` property which is part of the `GraphQLError` type. Please note that `code` is actually not part of it.

### CODE REVIEW

File: packages/application-shell/src/apollo-links/error-link.spec.js [hunk 1]
```diff
@@ -66,48 +66,102 @@ describe('with unauthenticated error', () => {
           },
         ],
       },
+      unauthenticatedWithStatusCode: {
+        data: null,
+        statusCode: 401,
+        errors: [
+          {
+            message: 'invalid_token',
```
Comment (tdeekens) [inline packages/application-shell/src/apollo-links/error-link.spec.js:74]: The `code` is not standard so we should just use message I suppose.

File: packages/application-shell/src/apollo-links/error-link.spec.js [hunk 2]
```diff
@@ -66,48 +66,102 @@ describe('with unauthenticated error', () => {
...
     };
     describe('when token retry is not skipped', () => {
-      beforeEach(async () => {
-        const debugLink = new ApolloLink((operation, forward) => {
-          context = operation.getContext();
-          operation.setContext({
-            headers: {
-              'X-Graphql-Target': GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM,
-            },
+      describe('when unauthorized via extension code', () => {
```
Comment (tdeekens) [inline packages/application-shell/src/apollo-links/error-link.spec.js:80]: The old test.

File: packages/application-shell/src/apollo-links/error-link.spec.js [hunk 3]
```diff
@@ -66,48 +66,102 @@ describe('with unauthenticated error', () => {
...
+              'X-Force-Token': 'true',
+            })
+          );
+        });
+        it('should eventually resolve with data', () => {
+          const [{ values }] = resolvedResponse;
+          expect(values).toEqual([responses.success]);
+        });
       });
+      describe('when unauthorized via status code', () => {
```
Comment (tdeekens) [inline packages/application-shell/src/apollo-links/error-link.spec.js:123]: The new test.

File: packages/application-shell/src/apollo-links/error-link.ts [hunk 4]
```diff
@@ -36,8 +36,12 @@ const errorLink = onError(
       const context = operation.getContext() as TApolloContext;
 
       for (const err of graphQLErrors) {
+        const isNonAuthenticatedViaExtensionCode =
+          err?.extensions?.code === 'UNAUTHENTICATED';
+        const isNonAuthenticatedViaCode = err?.message === 'invalid_token';
```
Comment (tdeekens) [inline packages/application-shell/src/apollo-links/error-link.ts:41]: Just moved up for clarity.

File: packages/application-shell/src/apollo-links/utils.ts [hunk 5]
```diff
@@ -60,7 +60,11 @@ const isGraphQLError = (
   error: ErrorResponse['graphQLErrors']
 ): error is GraphQLError[] =>
   Array.isArray(error) &&
-  error.some((err) => (err as GraphQLError).extensions !== undefined);
+  error.some(
+    (err) =>
+      (err as GraphQLError).message !== undefined ||
+      (err as GraphQLError).extensions !== undefined
```
Comment (tdeekens) [inline packages/application-shell/src/apollo-links/utils.ts:66]: I think message is even required on any `GraphQLError` according to the type. Maybe it alone is enough. Open for a discussion.

File: packages/application-shell/src/apollo-links/error-link.ts (hunk 4 above)
Comment (emmenko) [inline packages/application-shell/src/apollo-links/error-link.ts:41]: Should we mention here that this is a "workaround"? At least for us for future references or something.

File: packages/application-shell/src/apollo-links/error-link.spec.js [hunk 6]
```diff
@@ -66,48 +66,102 @@ describe('with unauthenticated error', () => {
           },
         ],
       },
+      unauthenticatedWithStatusCode: {
+        data: null,
+        statusCode: 401,
```
Comment (emmenko) [inline packages/application-shell/src/apollo-links/error-link.spec.js:71]: Is this correct?

**Review by emmenko — APPROVED** [review]
Thanks, looks good  and it should be "good enough" for now.

File: packages/application-shell/src/apollo-links/error-link.spec.js (hunk 6 above)
Comment (tdeekens) [reply packages/application-shell/src/apollo-links/error-link.spec.js:71]: Yeah, but we can remove it. Some APIs add a statusCode property.

Comment (tdeekens) [reply packages/application-shell/src/apollo-links/error-link.spec.js]: Removed.

File: packages/application-shell/src/apollo-links/error-link.ts (hunk 4 above)
Comment (tdeekens) [reply packages/application-shell/src/apollo-links/error-link.ts:48]: Added a comment.

**Pull request merged**